pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -379,14 +378,25 @@ async def _needed_remote_artifacts(self, batch):
                 async for remote_artifact in sync_to_async_iterable(
                     content_artifact._remote_artifact_saver_ras
                 ):
-                    if remote_artifact.remote_id == d_artifact.remote.pk:
+                    if d_artifact.url == remote_artifact.url:
+                        break
+
+                    if d_artifact.remote.pk == remote_artifact.remote_id:
+                        key = f"{str(content_artifact.pk)}-{str(remote_artifact.remote_id)}"
```

Comment:
This could also be changed like so because of auto-string conversion of fstrings

```suggestion
                        key = f"{content_artifact.pk}-{remote_artifact.remote_id}"
```